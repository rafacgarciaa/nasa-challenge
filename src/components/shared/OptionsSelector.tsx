import { useEffect, useState } from "react";
import RangeSlider from 'react-bootstrap-range-slider';
import DatePicker from "react-datepicker";

import { CameraType, FilterMode } from "../../store/ui/types";
import { Spinner } from "./Spinner";

import "react-datepicker/dist/react-datepicker.css";

interface IProps {
  cameras: CameraType[];
  cameraSelected?: CameraType | null;
  earthDateSelected?: Date | null;
  solMartialDaySelected?: number | null;
  loading?: boolean;
  filterMode?: FilterMode;
  solMartialDayMaximum?: number;
  solMartialDayMinimum?: number;
  onSelectCamera?: (selected: CameraType) => void;
  onSelectEarthDate?: (selected: Date) => void;
  onSelectSolMartianDay?: (selected: number) => void;
}

export const OptionsSelector = (props: IProps) => {
    const items: CameraType[] = props.cameras.length > 0
        ? [...props.cameras, 'ALL']
        : [];
    const itemSelected = props.cameraSelected || 'ALL';
    const modeSelected = props.filterMode || 'earth-date';

    const [currentMode, setMode] = useState(modeSelected);
    const isSolMartianDay = currentMode === 'sol-martian-day';
    const isEarthDate = currentMode === 'earth-date';
    
    useEffect(() => { setMode(modeSelected); }, [modeSelected]);

    const daySelected = props.solMartialDaySelected;
    const solMartialDaySelected =
        (daySelected === undefined || daySelected === null)
            ? 0
            : daySelected;
    const [currentDay, setDay] = useState(solMartialDaySelected);
    
    const dateSelected = props.earthDateSelected || new Date();
    const solMartialDayMax = props.solMartialDayMaximum || 100;
    const solMartialDayMin = props.solMartialDayMinimum || 0;

    return (
        <div className="btn-toolbar mb-2 mb-md-0">
            <div className="btn-group me-2">
                {props.loading && <Spinner />}
                {!props.loading && items.map(curr =>
                    <button
                        type="button"
                        key={curr}
                        className={"btn btn-sm btn-outline-secondary" + (curr === itemSelected ? ' active' : '')}
                        onClick={() => props.onSelectCamera && props.onSelectCamera(curr)}>{curr}</button>)}
            </div>
            <div className="me-2">
                <div className="btn btn-sm btn-outline-secondary">
                    { isEarthDate && <>
                        Earth Date
                        <div>
                            <DatePicker
                                selected={dateSelected}
                                onChange={(date: Date) => props.onSelectEarthDate && props.onSelectEarthDate(date)}
                                dateFormat="dd/MM/yyyy"
                            />
                        </div>
                    </> }
                    { isSolMartianDay && <>
                        Martian Day: <span className="badge badge-light">{currentDay}</span>
                        <div style={{'margin': '3px 0'}}>
                            <RangeSlider
                                min={solMartialDayMin}
                                max={solMartialDayMax}
                                value={currentDay}
                                onChange={changeEvent => {
                                    const newDayValue = parseInt(changeEvent.target.value, 10);
                                    setDay(newDayValue);
                                }}
                                onAfterChange={changeEvent => {
                                    const newDayValue = parseInt(changeEvent.target.value, 10);
                                    props.onSelectSolMartianDay && props.onSelectSolMartianDay(newDayValue);
                                }}
                                tooltip="off"
                            />
                        </div>
                    </> }
                    <div className="dropdown">
                        <div
                            className="dropdown-toggle"
                            data-bs-toggle="dropdown"
                            aria-expanded="false"
                            id="filterMenuButton">
                            <span data-feather="calendar"></span>
                        </div>
                        <ul className="dropdown-menu" aria-labelledby="filterMenuButton">
                            <li>
                                <button
                                    type="button"
                                    className={"dropdown-item" + (isEarthDate ? ' active' : '')}
                                    onClick={() => setMode('earth-date')}>By Earth Date
                                </button>
                            </li>
                            <li>
                                <button
                                    type="button"
                                    className={"dropdown-item" + (isSolMartianDay ? ' active' : '')}
                                    onClick={() => setMode('sol-martian-day')}>By Martian Sol
                                </button>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    );
}