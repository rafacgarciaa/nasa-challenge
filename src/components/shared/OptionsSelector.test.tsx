import "@testing-library/jest-dom";
import { shallow } from 'enzyme';
import moment from "moment";

import { OptionsSelector } from './OptionsSelector';

test('should render the text "Loading..."', () => {
  const wrapper = shallow(
    <OptionsSelector
      cameras={['FHAZ', 'RHAZ', 'MAST']}
      earthDateSelected={moment('2021-12-14').toDate()}
    />
  );

  expect(wrapper).toMatchSnapshot();

  const buttons = wrapper.find('.btn-group button');
  expect(buttons.length).toBe(4);
});
