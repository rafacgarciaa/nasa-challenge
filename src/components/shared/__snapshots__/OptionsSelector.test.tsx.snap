// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the text "Loading..." 1`] = `
<div
  className="btn-toolbar mb-2 mb-md-0"
>
  <div
    className="btn-group me-2"
  >
    <button
      className="btn btn-sm btn-outline-secondary"
      key="FHAZ"
      onClick={[Function]}
      type="button"
    >
      FHAZ
    </button>
    <button
      className="btn btn-sm btn-outline-secondary"
      key="RHAZ"
      onClick={[Function]}
      type="button"
    >
      RHAZ
    </button>
    <button
      className="btn btn-sm btn-outline-secondary"
      key="MAST"
      onClick={[Function]}
      type="button"
    >
      MAST
    </button>
    <button
      className="btn btn-sm btn-outline-secondary active"
      key="ALL"
      onClick={[Function]}
      type="button"
    >
      ALL
    </button>
  </div>
  <div
    className="me-2"
  >
    <div
      className="btn btn-sm btn-outline-secondary"
    >
      Earth Date
      <div>
        <r
          allowSameDay={false}
          customTimeInput={null}
          dateFormat="dd/MM/yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          enableTabLoop={true}
          excludeScrollbar={true}
          focusSelectedMonth={false}
          monthsShown={1}
          nextMonthAriaLabel="Next Month"
          nextMonthButtonLabel="Next Month"
          nextYearAriaLabel="Next Year"
          nextYearButtonLabel="Next Year"
          onBlur={[Function]}
          onCalendarClose={[Function]}
          onCalendarOpen={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          previousMonthAriaLabel="Previous Month"
          previousMonthButtonLabel="Previous Month"
          previousYearAriaLabel="Previous Year"
          previousYearButtonLabel="Previous Year"
          readOnly={false}
          renderDayContents={[Function]}
          selected={2021-12-14T03:00:00.000Z}
          selectsDisabledDaysInRange={false}
          shouldCloseOnSelect={true}
          showFourColumnMonthYearPicker={false}
          showFullMonthYearPicker={false}
          showMonthYearPicker={false}
          showPopperArrow={true}
          showPreviousMonths={false}
          showQuarterYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          showTwoColumnMonthYearPicker={false}
          showYearPicker={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
          yearItemNumber={12}
        />
      </div>
      <div
        className="dropdown"
      >
        <div
          aria-expanded="false"
          className="dropdown-toggle"
          data-bs-toggle="dropdown"
          id="filterMenuButton"
        >
          <span
            data-feather="calendar"
          />
        </div>
        <ul
          aria-labelledby="filterMenuButton"
          className="dropdown-menu"
        >
          <li>
            <button
              className="dropdown-item active"
              onClick={[Function]}
              type="button"
            >
              By Earth Date
            </button>
          </li>
          <li>
            <button
              className="dropdown-item"
              onClick={[Function]}
              type="button"
            >
              By Martian Sol
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
